### PR TITLE
feat: Project projection

### PR DESCRIPTION
**Pull request by fgerschwiler**

### CODE REVIEW

File: internal/iam/repository/eventsourcing/model/iam_test.go [hunk 1]
```diff
@@ -25,31 +25,31 @@ func TestProjectRoleAppendEvent(t *testing.T) {
 		{
 			name: "append set up start event",
 			args: args{
-				event: &es_models.Event{AggregateID: "AggregateID", Sequence: 1, Type: IAMSetupStarted, ResourceOwner: "OrgID"},
+				event: &es_models.Event{AggregateID: "AggregateID", Sequence: 1, Type: IAMSetupStarted, ResourceOwner: "GrantedOrgID"},
```
Comment (adlerhurst) [inline internal/iam/repository/eventsourcing/model/iam_test.go:28]: why did this code change? the test doesn't validate resource owner

Comment (adlerhurst) [reply internal/iam/repository/eventsourcing/model/iam_test.go:28]: also in the test cases below

File: internal/project/repository/view/model/project_grant_member_test.go [hunk 2]
```diff
@@ -27,15 +27,15 @@ func TestGrantedProjectMemberAppendEvent(t *testing.T) {
 		{
 			name: "append added member event",
 			args: args{
-				event:  &es_models.Event{AggregateID: "AggregateID", Sequence: 1, Type: es_model.ProjectGrantMemberAdded, ResourceOwner: "OrgID", Data: mockProjectGrantMemberData(&es_model.ProjectGrantMember{GrantID: "ProjectGrantID", UserID: "UserID", Roles: pq.StringArray{"Role"}})},
+				event:  &es_models.Event{AggregateID: "AggregateID", Sequence: 1, Type: es_model.ProjectGrantMemberAdded, ResourceOwner: "GrantedOrgID", Data: mockProjectGrantMemberData(&es_model.ProjectGrantMember{GrantID: "ProjectGrantID", UserID: "UserID", Roles: pq.StringArray{"Role"}})},
```
Comment (adlerhurst) [inline internal/project/repository/view/model/project_grant_member_test.go:30]: Why did you change these 2 test cases? the test does not check resource owner

File: internal/project/repository/view/model/project_member_test.go [hunk 3]
```diff
@@ -27,15 +27,15 @@ func TestProjectMemberAppendEvent(t *testing.T) {
 		{
 			name: "append added member event",
 			args: args{
-				event:  &es_models.Event{AggregateID: "AggregateID", Sequence: 1, Type: es_model.ProjectMemberAdded, ResourceOwner: "OrgID", Data: mockProjectMemberData(&es_model.ProjectMember{UserID: "UserID", Roles: pq.StringArray{"Role"}})},
+				event:  &es_models.Event{AggregateID: "AggregateID", Sequence: 1, Type: es_model.ProjectMemberAdded, ResourceOwner: "GrantedOrgID", Data: mockProjectMemberData(&es_model.ProjectMember{UserID: "UserID", Roles: pq.StringArray{"Role"}})},
```
Comment (adlerhurst) [inline internal/project/repository/view/model/project_member_test.go:30]: why did you change these two test cases? the test does not validate the resource owner

File: internal/project/repository/view/model/project_role_test.go [hunk 4]
```diff
@@ -25,18 +25,18 @@ func TestProjectRoleAppendEvent(t *testing.T) {
 		{
 			name: "append added member event",
 			args: args{
-				event:  &es_models.Event{AggregateID: "AggregateID", Sequence: 1, Type: es_model.ProjectRoleAdded, ResourceOwner: "OrgID", Data: mockProjectRoleData(&es_model.ProjectRole{Key: "Key", DisplayName: "DisplayName", Group: "Group"})},
+				event:  &es_models.Event{AggregateID: "AggregateID", Sequence: 1, Type: es_model.ProjectRoleAdded, ResourceOwner: "GrantedOrgID", Data: mockProjectRoleData(&es_model.ProjectRole{Key: "Key", DisplayName: "DisplayName", Group: "Group"})},
```
Comment (adlerhurst) [inline internal/project/repository/view/model/project_role_test.go:28]: IMO is the resource owner of a project role is the resource owner not the granted org. The name of the test is wrong also

Comment (adlerhurst) [reply internal/project/repository/view/model/project_role_test.go:28]: same for the tests below

**Review by adlerhurst — CHANGES_REQUESTED** [review]

File: internal/iam/repository/eventsourcing/model/iam_test.go (hunk 1 above)
Comment (fgerschwiler) [reply internal/iam/repository/eventsourcing/model/iam_test.go:28]: I think this was a refactoring problem i didnt realize

File: internal/query/project_role.go [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+type ProjectRoleSearchQueries struct {
+	SearchRequest
+	Queries []SearchQuery
+}
+
+func NewProjectRoleProjectIDSearchQuery(method TextComparison, value string) (SearchQuery, error) {
+	return NewTextQuery(ProjectRoleColumnProjectID, value, method)
+}
+
+func NewProjectRoleResourceOwnerSearchQuery(method TextComparison, value string) (SearchQuery, error) {
```
Comment (adlerhurst) [inline internal/query/project_role.go:146]: equals?

File: internal/query/project_role.go [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+	return sq.Select(
+			ProjectRoleColumnProjectID.identifier(),
+			ProjectRoleColumnCreationDate.identifier(),
+			ProjectRoleColumnChangeDate.identifier(),
+			ProjectRoleColumnResourceOwner.identifier(),
+			ProjectRoleColumnSequence.identifier(),
+			ProjectRoleColumnKey.identifier(),
+			ProjectRoleColumnDisplayName.identifier(),
+			ProjectRoleColumnGroupName.identifier()).
+			From(projection.ProjectRoleProjectionTable).PlaceholderFormat(sq.Dollar),
```
Comment (adlerhurst) [inline internal/query/project_role.go:215]:
```suggestion
			From(projectRolesTable.identifier()).PlaceholderFormat(sq.Dollar),
```

File: internal/query/project_role.go [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+			ProjectRoleColumnProjectID.identifier(),
+			ProjectRoleColumnCreationDate.identifier(),
+			ProjectRoleColumnChangeDate.identifier(),
+			ProjectRoleColumnResourceOwner.identifier(),
+			ProjectRoleColumnSequence.identifier(),
+			ProjectRoleColumnKey.identifier(),
+			ProjectRoleColumnDisplayName.identifier(),
+			ProjectRoleColumnGroupName.identifier(),
+			"COUNT(*) OVER ()").
+			From(projection.ProjectRoleProjectionTable).PlaceholderFormat(sq.Dollar),
```
Comment (adlerhurst) [inline internal/query/project_role.go:249]:
```suggestion
			From(projectRolesTable.identifier()).PlaceholderFormat(sq.Dollar),
```

File: internal/query/projection/project.go [hunk 8]
```diff
@@ -51,94 +58,131 @@ func (p *ProjectProjection) reducers() []handler.AggregateReducer {
...
+	e, ok := event.(*project.ProjectChangeEvent)
+	if !ok {
+		logging.LogWithFields("HANDL-dk2iF", "seq", event.Sequence(), "expected", project.ProjectChangedType).Error("wrong event type")
+		return nil, errors.ThrowInvalidArgument(nil, "HANDL-s00Fs", "reduce.wrong.event.type")
+	}
+	if e.Name == nil && e.HasProjectCheck == nil && e.ProjectRoleAssertion == nil && e.ProjectRoleCheck == nil && e.PrivateLabelingSetting == nil {
 		return crdb.NewNoOpStatement(e), nil
 	}
 
+	columns := make([]handler.Column, 0)
```
Comment (adlerhurst) [inline internal/query/projection/project.go:111]:
```suggestion
	columns := make([]handler.Column, 0, 7)
```

File: internal/query/projection/project.go [hunk 9]
```diff
@@ -51,94 +58,131 @@ func (p *ProjectProjection) reducers() []handler.AggregateReducer {
...
+	columns = append(columns, handler.NewCol(ProjectColumnChangeDate, e.CreationDate()))
+	columns = append(columns, handler.NewCol(ProjectColumnSequence, e.Sequence()))
```
Comment (adlerhurst) [inline internal/query/projection/project.go:113]:
```suggestion
	columns = append(columns, handler.NewCol(ProjectColumnChangeDate, e.CreationDate()),
		handler.NewCol(ProjectColumnSequence, e.Sequence()))
```

**Review by adlerhurst — CHANGES_REQUESTED** [review]

**Review by adlerhurst — APPROVED** [review]
